src/hooks/useNotificationModal.ts

```diff
@@ -0,0 +1,204 @@
+    );
+  }, [extraFields]);
+
+  // reset Targets when subscription method changes
+  useEffect(() => {
+    setTarget({});
+  }, [selectedSubscriptionMethod]);
+
+  // handles populating an error messages error array for the extra fields
+  useEffect(() => {
+    if (!extraFields) {
+      setExtraFieldErrorMessages([]);
+      return;
+    }
+    const extraFieldInputValEntries = Object.entries(extraFieldInputVals);
+    const nextErrorMessages = extraFieldInputValEntries
+      .map(([fieldName, fieldVal]) => {
+        // get the validator for the extra field
+        const validator = get(
+          extraFields.find(({ key }) => key === fieldName),
+          "validator",
+          () => ""
+        );
+        // validator returns an error message or an empty string
+        return validator(fieldVal);
+      })
+      .filter((v) => v); // filter out empty strings
+    setExtraFieldErrorMessages(nextErrorMessages);
+  }, [extraFieldInputVals, extraFields, setExtraFieldErrorMessages]);
+
+  useEffect(() => {
+    const targetEntries = Object.entries(target);
+    // check that required fields exist and there are no extra field errors
+    if (
+      !targetEntries.length ||
+      !selectedTriggerId ||
+      !selectedSubscriptionMethod ||
+      extraFieldErrorMessages.length
+    ) {
+      setIsFormValid(false);
+      return;
+    }
+    // check that subscription methods input has the correct value
+    // targetEntries should only ever be length 1
+    const isTargetValid = targetEntries.reduce(
+      (accum, [tName, tVal]) =>
+        accum && subscriptionMethodControls[tName].validator(tVal),
+      true
+    );
+    setIsFormValid(isTargetValid);
+  }, [
+    subscriptionMethodControls,
+    target,
+    extraFieldInputVals,
+    selectedTriggerId,
+    selectedSubscriptionMethod,
+    extraFieldErrorMessages,
+  ]);
+
+  const getRequestPayload = () => {
+    const targetEntry = Object.entries(target)[0];
+    return {
+      trigger: selectedTriggerId,
+      resource_type: resourceType,
+      selectors: [
+        { type: "object", data: resourceType.toLowerCase() },
+        { type: "id", data: resourceId },
+      ],
+      subscriber: {
+        type: targetEntry[0],
+        target: targetEntry[1],
+      },
+      trigger_data: Object.entries(extraFieldInputVals).reduce(
+        (accum, [key, val]) => ({ ...accum, [key]: `${val}` }),
+        {}
+      ),
+      owner_type: "person",
+      regex_selectors: [],
+    };
+  };
+
+  const onClickSave = () => {
+    saveSubscription({
+      variables: { subscription: getRequestPayload() },
+    });
+    onCancel();
+  };
+
+  return {
+    selectedSubscriptionMethod,
+    onClickSave,
+    mutationLoading,
+    isFormValid,
+    selectedTriggerId,
+    setSelectedTriggerId,
+    extraFields,
+    extraFieldInputVals,
+    setExtraFieldInputVals,
+    setSelectedSubscriptionMethod,
+    target,
+    setTarget,
+    extraFieldErrorMessages,
+  };
```

Comment:
How do you feel about organizing the exports of your hook, so that read only values are separate from functions that perform some type of mutation or action. Similarly to how the `usePatchStatusSelect` Hook is written or the `useMutation` and `useQuery` hooks are written.

Reply:
Something like `return [isFormValid, target, extraFieldInputVals, ..., {setTarget, setExtraFieldInputVals, ...}]`

Reply:
sure

Reply:
actually, I'm unsure if the extra level of organization will be helpful. The example you posted exports the variables in an array but I don't think order is important here. Also I think the variable names make it easy to understand which variables are functions.